From 80997b4480cc00f1b908aa29258751224d7911be Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 22 Jan 2024 20:26:20 +0100
Subject: [PATCH 550/954] feat(stm32mp2): add DDR3 platform power functions

Add the required power function ddr_power_init() for STM32MP_DDR3_TYPE

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 32a49c6b6c727249dbf2f87462a87a724844a9dd feat(stm32mp2): add DDR3 platform power functions]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Ia7dd661e008b194522b96ff75a7e33652bdee9b4
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/355243
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/368534
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 plat/st/stm32mp2/plat_ddr.c | 51 ++++++++++++++++++++++++++++++++++++-
 1 file changed, 50 insertions(+), 1 deletion(-)

diff --git a/plat/st/stm32mp2/plat_ddr.c b/plat/st/stm32mp2/plat_ddr.c
index 05923e04c5..60419c8166 100644
--- a/plat/st/stm32mp2/plat_ddr.c
+++ b/plat/st/stm32mp2/plat_ddr.c
@@ -16,6 +16,53 @@
 
 #include <platform_def.h>
 
+#if STM32MP_DDR3_TYPE
+struct ddr3_supply {
+	struct rdev *vdd;
+	struct rdev *vref;
+	struct rdev *vtt;
+};
+
+static void ddr3_supply_read(void *fdt, int node, struct ddr3_supply *supply)
+{
+	supply->vdd = regulator_get_by_supply_name(fdt, node, "vdd");
+	supply->vref = regulator_get_by_supply_name(fdt, node, "vref");
+	supply->vtt = regulator_get_by_supply_name(fdt, node, "vtt");
+}
+
+static int ddr_power_init(void *fdt, int node)
+{
+	int status;
+	struct ddr3_supply supply;
+
+	ddr3_supply_read(fdt, node, &supply);
+	if ((supply.vdd == NULL) || (supply.vref == NULL) || (supply.vtt == NULL)) {
+		return -ENOENT;
+	}
+
+	/*
+	 * DDR3 power on sequence is:
+	 * enable VREF_DDR, VTT_DDR, VPP_DDR
+	 */
+	status = regulator_set_min_voltage(supply.vdd);
+	if (status != 0) {
+		return status;
+	}
+
+	status = regulator_enable(supply.vdd);
+	if (status != 0) {
+		return status;
+	}
+
+	status = regulator_enable(supply.vref);
+	if (status != 0) {
+		return status;
+	}
+
+	return regulator_enable(supply.vtt);
+}
+#endif /* STM32MP_DDR3_TYPE */
+
 #if STM32MP_DDR4_TYPE
 struct ddr4_supply {
 	struct rdev *vdd;
@@ -150,7 +197,9 @@ int stm32mp_board_ddr_power_init(enum ddr_type ddr_type)
 
 	VERBOSE("DDR power init, ddr_type = %u\n", ddr_type);
 
-#if STM32MP_DDR4_TYPE
+#if STM32MP_DDR3_TYPE
+	assert(ddr_type == STM32MP_DDR3);
+#elif STM32MP_DDR4_TYPE
 	assert(ddr_type == STM32MP_DDR4);
 #elif STM32MP_LPDDR4_TYPE
 	assert(ddr_type == STM32MP_LPDDR4);
-- 
2.53.0

